### DISCUSSION TITLE
R_SPF_FAIL for a vaild SPF

### DISCUSSION THREAD

**Discussion by mkrinkl** [rspamd/rspamd#3963] [Q&A]
Hello,

Recently we started having certain email quarantined although they have a valid SPF record.
The spam score is being attributed based on SPF (and other things but with SPF having the highest value) and emails quarantined.
We are using the rspamd module with mailcow-dockerized.

At this point I am not sure where to go next with the investigation and hoping for some pointer from someone that might have experienced this or similar issue before.

So here’s the mailcow quarantine screenshot
![quarantine](https://user-images.githubusercontent.com/83456409/140555484-8b415a4e-bd12-4c5a-a13e-26818df6526f.png)

But the domain in question seems tho have a perfectly valid SPF record:

```
<<>> DiG 9.16.15-Debian <<>> txt wasdkeyboards.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 14215
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;wasdkeyboards.com. IN TXT

;; ANSWER SECTION:
wasdkeyboards.com. 300 IN TXT “v=spf1 include:spf.mandrillapp.com include:_spf.google.com include:mail.zendesk.com -all”

;; Query time: 24 msec
;; SERVER: 10.24.1.1#53(10.24.1.1)
;; WHEN: Fri Nov 05 12:03:04 CET 2021
;; MSG SIZE rcvd: 136
```

I would appreciate anyone that had this type of issue taking a look and suggesting what/where to look.
I am not interested in whitelisting/raising/lowering the R_SPF_FAIL at this point.
I would like to be able to understand why it assigns a high score to a vaild SPF.

Thanks,

**Comment by vstakhov** [comment]
SPF_FAIL means SPF rejection not an invalid record. I cannot say more without knowledge of the source IP.

**Comment by mkrinkl** [comment]
Hello,

Thank you for your reply. Apologies or the delay in providing more info.

So, I can't get that particular IP at this moment, but here#s another example of the same issue.
![projectbike lu](https://user-images.githubusercontent.com/83456409/142441258-8f8e6d4f-91ee-4c70-ab81-d5a7dd1ad5dc.png)

The source IP in this example is 198.2.131.225
It seems to be clean and not blacklisted anywhere.

Here's the mailcow-postfix log:
`rspamd-mailcow_1     | 2021-11-17 10:08:06 #40(normal) <216e26>; task; rspamd_task_write_log: id: <c77bb1dc328b7c1e1b8672260.332f3f1c90.20211117085923.6bc26fafee.f1e512ae@mail225.atl121.mcsv.net>, qid: <AFED73001F8>, ip: 31.22.124.147, from: <bounce-mc.us8_28613243.5400254-332f3f1c90@mail225.atl121.mcsv.net>, (default: T (reject): [25.20/15.00] [R_SPF_FAIL(8.00){-all:c;},IP_REPUTATION_SPAM(5.12){asn: 197692(0.34), country: LU(0.00), ip: 31.22.124.147(0.94);},BAYES_SPAM(4.50){100.00%;},NEURAL_HAM_LONG(-4.00){-1.000;},VIOLATED_DIRECT_SPF(3.50){},FORGED_W_BAD_POLICY(3.00){},BAD_REP_POLICIES(2.00){},NEURAL_HAM_SHORT(-2.00){-1.000;},URIBL_GREY(1.50){list-manage.com:url;eepurl.com:url;},FROM_EXCESS_QP(1.20){},REPLYTO_EXCESS_QP(1.20){},URI_COUNT_ODD(1.00){37;},BAD_WORDS(0.20){},MIME_GOOD(-0.10){multipart/alternative;text/plain;},RCVD_NO_TLS_LAST(0.10){},HAS_LIST_UNSUB(-0.01){},MX_GOOD(-0.01){},ARC_NA(0.00){},ARC_SIGNED(0.00){example.coms=dkim:i=1;},ASN(0.00){asn:197692, ipnet:31.22.120.0/21, country:LU;},BCC(0.00){},DKIM_TRACE(0.00){mailchimpapp.net:+;},DMARC_NA(0.00){projectbike.lu;},FORGED_SENDER(0.00){info@projectbike.lu;bounce-mc.us8_28613243.5400254-332f3f1c90@mail225.atl121.mcsv.net;},FROM_HAS_DN(0.00){},FROM_NEQ_ENVFROM(0.00){info@projectbike.lu;bounce-mc.us8_28613243.5400254-332f3f1c90@mail225.atl121.mcsv.net;},HAS_REPLYTO(0.00){info@projectbike.lu;},MIME_TRACE(0.00){0:+;1:+;2:~;},PREVIOUSLY_DELIVERED(0.00){user@example.com;},RCPT_COUNT_ONE(0.00){1;},RCPT_MAILCOW_DOMAIN(0.00){example.com;},RCVD_COUNT_THREE(0.00){3;},RCVD_IN_DNSWL_NONE(0.00){198.2.131.225:received;},REDIRECTOR_FALSE(0.00){projectbike.lu->list-manage.com;eepurl->list-manage.com;},REPLYTO_EQ_FROM(0.00){},R_DKIM_ALLOW(0.00){mailchimpapp.net:s=k2;},TO_DN_NONE(0.00){},TO_MATCH_ENVRCPT_ALL(0.00){}]), len: 67723, time: 1475.131ms, dns req: 63, digest: <6e56b726ecbf2071719dd57431f1b185>, rcpts: <user@example.com>, mime_rcpts:`

Thank you for taking the time to check this. If any other logs or config settings are needed, I'll be happy to provide it.

Thanks!

**Reply by kwisatz** [reply]
@vstakhov Joining in here, working at the same company as @mkrinkl.
From the above record, we can see both an `R_SPF_FAIL` and a `VIOLATED_DIRECT_SPF`.
However, the domain in question does not provide any SPF:

```bash
kwisatz@thufir:~$ dig +short TXT projectbike.lu
"1|www.projectbike.lu"
kwisatz@thufir:~$ dig +short SPF projectbike.lu
kwisatz@thufir:~$ 
```

How can we get more verbose logs on SPF evaluation to see why, in this case, a non-existing record leads to those results?

**Comment by mkrinkl** [comment]
Hi @vstakhov,

Any chance we can get your opinion on this? That would be greatly appreciated.

Thanks,

**Reply by dupondje** [reply]
Afaik you also need to check SPF record of mail225.atl121.mcsv.net.
mail225.atl121.mcsv.net. 86400	IN	TXT	"v=spf1 ip4:198.2.131.225 include:spf.mandrillapp.com -all"


this seems to contain the IP now, but of course it could have been added after the report.

**Comment by kwisatz** [comment]
Bumping this once more with additional information and a new example:

Here's an email where SPF should have failed but it didn't. Full headers with truncated recipient/faked sender:

```
Received: from vm3315861.24ssd.had.wf (localhost [127.0.0.1])
	by vm3315861.24ssd.had.wf (8.14.7/8.14.7) with ESMTP id 228I1YpU028785
	for <$USER@mga.lu>; Tue, 8 Mar 2022 20:01:34 +0200
Received: (from root@localhost)
	by vm3315861.24ssd.had.wf (8.14.7/8.14.7/Submit) id 228I1Ysc028765;
	Tue, 8 Mar 2022 20:01:34 +0200
Received: from mail.mymx.lu ([172.22.1.253])
	by 21ac8f7e1993 with LMTP
	id SMCnHQSaJ2LgZQAAlX/YNQ
	(envelope-from <root@vm3315861.24ssd.had.wf>)
	for <$USER@mga.lu>; Tue, 08 Mar 2022 19:01:40 +0100
Received: from vm3315861.24ssd.had.wf (vm3315861.24ssd.had.wf [80.89.229.175])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by mail.mymx.lu (Postcow) with ESMTPS id A8964305628
	for <$USER@mga.lu>; Tue,  8 Mar 2022 19:01:39 +0100 (CET)
From: "fax@mga.lu" <$USER@mga.lu>
To: <$USER@mga.lu>
Subject: =?UTF-8?Q?=E2=9C=85_Notification:_You_have_receive?=
	=?UTF-8?Q?d_a_new_fax_document_-_4_page=28s=29?=
Date: Tue, 8 Mar 2022 19:01:34 +0100
Message-ID: <202203081801.228I1Ysc028765@vm3315861.24ssd.had.wf>
MIME-Version: 1.0
Content-Type: multipart/alternative;
	boundary="----=_NextPart_000_0076_01D833C1.FEAD8900"
X-Last-TLS-Session-Version: TLSv1.2
X-Rspamd-Queue-Id: A8964305628
Authentication-Results: mail.mymx.lu;
	dkim=none;
	dmarc=none;
	spf=none (mail.mymx.lu: domain of root@vm3315861.24ssd.had.wf has no SPF policy when checking 80.89.229.175) smtp.mailfrom=root@vm3315861.24ssd.had.wf
X-Spamd-Result: default: False [7.33 / 15.00];
	BAYES_SPAM(4.50)[99.99%];
	NEURAL_HAM_LONG(-4.00)[-1.000];
	MX_MISSING(3.50)[];
	FORGED_W_BAD_POLICY(3.00)[];
	NEURAL_HAM_SHORT(-2.00)[-1.000];
	AUTH_NA(1.00)[];
	MV_CASE(0.50)[];
	MX_INVALID(0.50)[];
	MIME_HTML_ONLY(0.20)[];
	MIME_BASE64_TEXT(0.10)[];
	IP_REPUTATION_SPAM(0.03)[asn: 204601(0.00), country: NL(0.01), ip: 80.89.229.175(0.00)];
	TO_MATCH_ENVRCPT_ALL(0.00)[];
	PREVIOUSLY_DELIVERED(0.00)[$USER@mga.lu];
	MAILCOW_DOMAIN_HEADER_FROM(0.00)[mga.lu];
	RCPT_COUNT_ONE(0.00)[1];
	DMARC_NA(0.00)[mga.lu];
	R_SPF_NA(0.00)[no SPF record];
	FROM_HAS_DN(0.00)[];
	BCC(0.00)[];
	ASN(0.00)[asn:204601, ipnet:80.89.228.0/23, country:NL];
	ARC_NA(0.00)[];
	RCVD_COUNT_THREE(0.00)[3];
	FORGED_SENDER(0.00)[$USER@mga.lu,root@vm3315861.24ssd.had.wf];
	ARC_SIGNED(0.00)[mga.lu:s=dkim:i=1];
	TO_DN_NONE(0.00)[];
	HAS_DATA_URI(0.00)[];
	RCPT_MAILCOW_DOMAIN(0.00)[mga.lu];
	R_DKIM_NA(0.00)[];
	FROM_NEQ_ENVFROM(0.00)[$USER@mga.lu,root@vm3315861.24ssd.had.wf];
	RCVD_TLS_LAST(0.00)[];
	MIME_TRACE(0.00)[0:~];
	TO_EQ_FROM(0.00)[];
	GREYLIST(0.00)[pass,body]
Thread-Index: AQEJErZWwdpC62BIOKr5a7IHKqfbBg==

This is a multipart message in MIME format.
```

The most interesting part here is this one:

```
Authentication-Results: mail.mymx.lu;
	dkim=none;
	dmarc=none;
	spf=none (mail.mymx.lu: domain of root@vm3315861.24ssd.had.wf has no SPF policy when checking 80.89.229.175) smtp.mailfrom=root@vm3315861.24ssd.had.wf
```

There is no other mention of SPF. So here, only the originating server/envelope from was checked for SPF, but not the header from server which should IMHO at least also have been checked (If not, SPF is useless).

For `mga.lu`, an SPF record exists:

```
kwisatz@thufir:~$ dig +short TXT mga.lu
"v=spf1 a mx ~all"
```

Is this a configuration problem on our or mailcow's side or is it an rspamd bug?

**Reply by vstakhov** [reply]
> There is no other mention of SPF. So here, only the originating server/envelope from was checked for SPF, but not the header from server which should IMHO at least also have been checked (If not, SPF is useless).

Please read the SPF standard. It has nothing related to a message's headers.

**Reply by kwisatz** [reply]
I'm sorry, I don't understand that. If that's true (and I will go and read the RFC later), then why would e.g. mailchimp for instance have instructions for including their SPF rules when all they do is set a `from` header ?

E.g. 

```
(envelope-from <bounce-mc.us11_42920509.459682-08e5a01d61@mail9.us4.mcsv.net>)
…
From: =?utf-8?Q?WASD=20Keyboards?= <noreply@wasdkeyboards.com>
```

And the SPF:

```
wasdkeyboards.com.	300	IN	TXT	"v=spf1 include:spf.mandrillapp.com include:_spf.google.com include:mail.zendesk.com -all"
```


If what you're saying is correct, then the only case where SPF would apply and penalize me would be if I tried to set the domain of the server that I'm sending from to a value for which an SPF record exists?

**Reply by kwisatz** [reply]
Ok, you're absolutely right. There's even [a section for this on the Wikipedia](https://en.wikipedia.org/wiki/Sender_Policy_Framework#Header_limitations), I didn't have to go and read the RFC.
And that means that setting this include for mailchimp/mandrill is actually useless.

**Comment by kwisatz** [comment]
If we disregard my blunder with the `from` header and cycle around to my initial post, I have followed up one of those mails in more detail.

E.g. an envelope-from of `bounce-mc.us4_8716717.10093822-c7c0a81160@mail247.suw16.rsgsv.net` that got an `R_SPF_FAIL (8)`.
We have

```
mail247.suw16.rsgsv.net. 86400	IN	TXT	"v=spf1 ip4:198.2.182.247 include:spf.mandrillapp.com -all"

spf.mandrillapp.com.	13766	IN	TXT	"v=spf1 ip4:198.2.128.0/24 ip4:198.2.132.0/22 ip4:198.2.136.0/23 ip4:198.2.145.0/24 ip4:198.2.186.0/23 ip4:205.201.131.128/25 ip4:205.201.134.128/25 ip4:205.201.136.0/23 ip4:205.201.139.0/24 ip4:198.2.177.0/24 ip4:198.2.178.0/23 ip4:198.2.180.0/24 ~all"
```

and indeed, that list does not include `mail247.suw16.rsgsv.net`:

````
mail247.suw16.rsgsv.net. 86400	IN	A	198.2.182.247
```
